internal/pkg/gateway/api.go

```diff
@@ -273,6 +270,43 @@ func (gs *Server) Endorse(ctx context.Context, request *gp.EndorseRequest) (*gp.
 	return endorseResponse, nil
 }
 
+// responseStatus unpacks the proposal response and error values that are returned from ProcessProposal and
+// determines how the gateway should react (retry?, close connection?).
+// Uses the grpc canonical status error codes and their recommended actions.
+// Returns:
+// - response successful (bool)
+// - error message extracted from the err or generated from 500 proposal response
+// - should the gateway retry (only the Evaluate() uses this)
+// - should the gateway close the connection and remove the peer from its registry
+func (gs *Server) responseStatus(response *peer.ProposalResponse, err error) (bool, string, bool, bool) {
```

Comment:
```suggestion
func (gs *Server) responseStatus(response *peer.ProposalResponse, err error) (success bool, errorMessage string, retry bool, remove bool) {
```

These functions that return multiple bools can be difficult to keep track of. Adding variable names in the function declaration helps the reader keep track of things and tie them back to the comments above, even if they are not used in the body.

Reply:
I'm not a big fan of long lists of parameters or return values in general. I wonder if encapsulating the logic and/or resulting information in a separate struct (with nicely named methods, or at least fields) to communicate information back to consuming code might be clearer?

It's a shame that this logic is required to distinguish between different classes of errors at all though. 😞